File: student_Banasthali_Nivedita/graphs/problem13.cpp
/*
Given a list of airline tickets represented as pairs of origin and destination airports, implement a function to
reconstruct the itinerary in order, assuming that the itinerary starts from "JFK".

*/

#include<bits/stdc++.h>

using namespace std;

void dfs(string& currentAirport, unordered_map<string, vector<string>>& flights, vector<string>& itinerary) {
    vector<string>& destinations = flights[currentAirport];

    while (!destinations.empty()) {
        string nextAirport = destinations.back();
        destinations.pop_back();
        dfs(nextAirport, flights, itinerary);
    }

    itinerary.push_back(currentAirport);
}

vector<string> reconstructItinerary(vector<vector<string>>& tickets) {
    unordered_map<string, vector<string>> flights;
    vector<string> itinerary;

    // Build the graph of flights
    for (const vector<string>& ticket : tickets) {
        string from = ticket[0];
        string to = ticket[1];
        flights[from].push_back(to);
    }

    // Sort the destinations in lexical order
    for (auto& entry : flights) {
        sort(entry.second.begin(), entry.second.end(), greater<string>());
    }

    // Reconstruct the itinerary starting from "JFK"
    dfs("JFK", flights, itinerary);

    reverse(itinerary.begin(), itinerary.end());
    return itinerary;
}

int main() {
    vector<vector<string>> tickets = {
        {"MUC", "LHR"},
        {"JFK", "MUC"},
        {"SFO", "SJC"},
        {"LHR", "SFO"}
    };

    vector<string> itinerary = reconstructItinerary(tickets);
    cout << "Reconstructed itinerary: ";
    for (const string& airport : itinerary) {
        cout << airport << " ";
    }
    cout << endl;

    return 0;
}

File: ShrutiVarshney_BanasthaliVidyapith_GraphsAssignment/GraphsAssignmentQuestions/src/graphsassignmentquestions/Problem13.java
package graphsassignmentquestions;
import java.util.*;
public class Problem13 
{
    public List<String> findItinerary(List<List<String>> tickets) 
    {
        Map<String, PriorityQueue<String>> graph = new HashMap<>();
        List<String> itinerary = new ArrayList<>();

        // Build the graph
        for (List<String> ticket : tickets) 
        {
            String from = ticket.get(0);
            String to = ticket.get(1);
            graph.putIfAbsent(from, new PriorityQueue<>());
            graph.get(from).offer(to);
        }

        // Start DFS from "JFK"
        dfs("JFK", graph, itinerary);

        // Reverse the itinerary to get the correct order
        Collections.reverse(itinerary);

        return itinerary;
    }

    private void dfs(String airport, Map<String, PriorityQueue<String>> graph, List<String> itinerary) 
    {
        PriorityQueue<String> destinations = graph.get(airport);

        while (destinations != null && !destinations.isEmpty())
        {
            String nextAirport = destinations.poll();
            dfs(nextAirport, graph, itinerary);
        }

        itinerary.add(airport);
    }

    public static void main(String[] args) 
    {
        List<List<String>> tickets = new ArrayList<>();
        tickets.add(Arrays.asList("MUC", "LHR"));
        tickets.add(Arrays.asList("JFK", "MUC"));
        tickets.add(Arrays.asList("SFO", "SJC"));
        tickets.add(Arrays.asList("LHR", "SFO"));

        Problem13 solution = new Problem13();
        List<String> itinerary = solution.findItinerary(tickets);
        System.out.println(itinerary); // Output: [JFK, MUC, LHR, SFO, SJC]
    }
}
